addons/mail/controllers/gif.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request, route, Controller
+import requests
+import json
+
+
+class MailGifController(Controller):
+
+    ckey = "my_test_app"
+    limit = 8
+    # This is the equivalent of PG-13
+    # see https://developers.google.com/tenor/guides/content-filtering
+    contentFilter = "low"
+    tenor_url = "https://tenor.googleapis.com"
+
+    @route("/discuss/gif/search", type="json", auth="public", cors="*")
+    def mail_gif_search(self, search_term, locale="en", country="US", pos=False):
+        api_key = request.env["ir.config_parameter"].sudo().get_param("mail.tenor_api_key")
+        url = f"{self.tenor_url}/v2/search?q={search_term}&key={api_key}&client_key={self.ckey}&limit={self.limit}&contentfilter={self.contentFilter}&locale={locale}&country={country}&media_filter=tinygif"
+        if pos:
+            url = f"{url}&pos={pos}"
+        r = requests.get(url, timeout=3)
+        return json.loads(r.content)
+
+    @route("/discuss/gif/categories", type="json", auth="public", cors="*")
+    def mail_gif_categorie(self, locale="en", country="US"):
+        api_key = request.env["ir.config_parameter"].sudo().get_param("mail.tenor_api_key")
+        url = f"{self.tenor_url}/v2/categories?key={api_key}&client_key={self.ckey}&contentfilter={self.contentFilter}&locale={locale}&country={country}"
+        r = requests.get(url, timeout=3)
+        return json.loads(r.content)
+
+    @route("/discuss/gif/set_favorite", type="json", auth="user")
+    def set_gif_favorite(self, tenor_gif_id):
+        request.env["mail.gif.favorite"].create({"tenor_gif_id": tenor_gif_id})
+
+    def mail_gif_posts(self, ids):
+        api_key = request.env["ir.config_parameter"].sudo().get_param("mail.tenor_api_key")
+        url = f"{self.tenor_url}/v2/posts?key={api_key}&client_key={self.ckey}&ids={','.join(ids)}&media_filter=tinygif"
+        r = requests.get(url, timeout=3)
+        return json.loads(r.content)["results"]
+
+    @route("/discuss/gif/favorites", type="json", auth="user")
+    def get_gif_favorites(self, offset=0):
+        tenor_gif_ids = request.env["mail.gif.favorite"].search([("create_uid", "=", request.env.user.id)], limit=20, offset=offset)
+        return {
+            "results": self.mail_gif_posts(tenor_gif_ids.mapped("tenor_gif_id")) or [],
+            "offset": offset + 20
+        }
+
+    @route("/discuss/gif/remove_favorite", type="json", auth="user")
+    def remove_gif_favorite(self, tenor_gif_id):
+        return request.env["mail.gif.favorite"].search([("create_uid", "=", request.env.user.id), ("tenor_gif_id", "=", tenor_gif_id)]).unlink()
```

Comment:
Don't we need to make some kind of rate limiting with their API?

Reply:
I don't think so, they handle the rate limit on their side and I have never reach it at all, even when I tried to.
They say request just fail when above the tenor threshold and that is now handled in the code. Once you are above the threshold (it's 1sec so, it's not failing for long), the API work's fine.

https://developers.google.com/tenor/guides/rate-limits-and-caching